lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,386 @@
+            this.restorer = null;
+            this.topics = backendConfig.topics;
+        }
+
+
+        /**
+         * Queries the database to get a random record. Returns that random
+         * record along with methods to handle it.
+         */
+        checkNow() {
+            var wrappedMessage;
+
+            this.connection.query(`SELECT *
+                FROM ??
+                WHERE topics IN (?)
+                OR topics IS NULL
+                ORDER BY RAND()
+                LIMIT 1;`,
+                [
+                    this.config.table,
+                    this.config.topics
+                ],
+                (err, record) => {
+                    if (err) {
+                        this.emit(EMIT_ERR, err);
+                    } else {
+                        wrappedMessage = {
+                            heartbeat: heartbeat.bind(null, this, record[0].hash),
+                            message: record[0].messageBase64,
+                            requeue: requeue.bind(null, this, record[0].hash),
+                            remove: remove.bind(null, this, record[0].hash),
+                            topics: record[0].topics
+                        };
+
+                        this.emit(EMIT_DATA, wrappedMessage);
+                    }
+                }
+            );
+        }
+
+
+        /**
+         * Opens a connection to a MySQL database.
+         *
+         * @param {Function} callback
+         */
+        connect(callback) {
+            var connectionOptions, mysqlOptions;
+
+            /*
+            These are not all of the options that a mysql connection can
+            take, only what is most relevant to this plugin.
+            See the MySQL documentation for the rest.
+            */
+            mysqlOptions = [
+                "database",
+                "host",
+                "password",
+                "port",
+                "table",
+                "user"
+            ];
+            connectionOptions = {};
+
+            mysqlOptions.forEach((key) => {
+                if (this.config[key]) {
+                    connectionOptions[key] = this.config[key];
+                }
+            });
+
+            this.connection = mysql.createConnection(connectionOptions);
+            this.connection.connect((err) => {
+                if (err) {
+                    throw new Error("There was an error while attempting to connect to the database.");
+                }
+
+                callback(false);
+            });
+        }
+
+
+        /**
+         * Severs the connection from the database.
+         *
+         * @param {Function} callback
+         */
+        disconnect(callback) {
+            this.connection.end(callback);
+        }
+
+
+        /**
+         * Starts polling.
+         *
+         * @param {Function} callback
+         */
+        listen(callback) {
+            this.currentlyPolling = true;
+            poll(this, callback);
```

Comment:
The callback passed here is used to signal when you finish the listen task.  You set up polls and restores.  The way you pass the callback to poll means the callback is called whenever you perform a poll.  The callback will be called thousands of times, not just once.  It's not an "on completion" callback.

I'd drop the callback here and not pass it to poll.  Then, at the end of this function I would call the callback synchronously.  Perhaps an easier solution would be to not use a callback here at all, similar to pausePolling, resumePolling, etc.